client/src/app/primitives/TabLinks.js

```diff
@@ -147,57 +105,139 @@ export default class TabLinks extends PureComponent {
           {
             tabs.map(tab => {
               const dirty = this.isDirty(tab);
+              const active = tab === activeTab;
 
               return (
-                <span
+                <Tab
                   key={ tab.id }
-                  data-tab-id={ tab.id }
-                  title={ tab.title }
-                  className={ classNames('tab', {
-                    active: tab === activeTab,
-                    dirty
-                  }) }
-                  onClick={ () => onSelect(tab, event) }
-                  onContextMenu={ (event) => (onContextMenu || noop)(tab, event) }
-                  draggable
-                >
-                  {tab.name}
-                  {
-                    onClose && <span
-                      className="close"
-                      title="Close Tab"
-                      onClick={ e => {
-                        e.preventDefault();
-                        e.stopPropagation();
-
-                        onClose(tab);
-                      } }
-                    >
-                      {
-                        dirty ? <CircleIcon className="icon dirty-icon" /> : null
-                      }
-                      <TabCloseIcon className="icon close-icon" />
-                    </span>
-                  }
-                </span>
+                  tab={ tab }
+                  active={ active }
+                  dirty={ dirty }
+                  getTabIcon={ getTabIcon }
+                  onClose={ onClose }
+                  onContextMenu={ onContextMenu }
+                  onSelect={ onSelect }
+                />
               );
             })
           }
 
           {
-            placeholder && <span
+            placeholder && <div
               key="__placeholder"
-              className={ classNames('tab placeholder ignore', {
-                active: tabs.length === 0
+              className={ classNames('tab tab--ignore tab--placeholder', {
+                'tab--active': tabs.length === 0
               }) }
               onClick={ placeholder.onClick }
               title={ placeholder.title }
             >
-              { placeholder.label }
-            </span>
+              <span className="tab__content">
+                { placeholder.label }
+              </span>
+            </div>
           }
         </div>
       </div>
     );
   }
 }
+
+function Tab(props) {
+  const {
+    active,
+    dirty,
+    getTabIcon,
+    onClose,
+    onContextMenu,
+    onSelect,
+    tab
+  } = props;
+
+  const tabRef = React.useRef(0);
+
+  // set tab to <small> once a defined threshold is reached (be responsive)
+  React.useEffect(() => {
+    if (!tabRef.current) {
+      return;
+    }
+
+    const tabNode = tabRef.current;
+    const width = tabNode.getBoundingClientRect().width;
+
+    tabNode.classList.toggle(TAB_SMALL_SELECTOR, width < SMALL_TAB_WIDTH);
+  });
```

Comment:
Solution sketch for https://github.com/camunda/camunda-modeler/pull/2589#issuecomment-984508557

```suggestion

  const tabNode = tabRef.current;

  // set tab to <small> once a defined threshold is reached (be responsive)
  React.useEffect(() => {
    if (!tabNode) {
      return;
    }

    const resizeObserver = new ResizeObserver(() => {
      const width = tabNode.getBoundingClientRect().width;

      tabNode.classList.toggle(TAB_SMALL_SELECTOR, width < SMALL_TAB_WIDTH);
    });

    resizeObserver.observe(tabNode);

    return () => resizeObserver.disconnect();
  }, [ tabNode ]);
```

Reply:
So instead of checking the width on render, we do this whenever node is resized.

Reply:
I think was is missing here is to react to the _initial size_ of a tab (first render). E.g. when I open 15+ tabs at once, it wouldn't react because there is no resizing included.

Reply:
Okay, actually the initial state would be respected as well. Instead, I'd probably handle the class via state, as we do with `dirty` and `active`. The reason is that it would override the classes on every render. Something like

```js
const [ small, setSmall ] = React.useState(false);

const tabNode = tabRef.current;

React.useEffect(() => {
  if (!tabNode) {
    return;
  }

  const resizeObserver = new ResizeObserver(() => {
    const width = tabNode.getBoundingClientRect().width;
    setSmall(width < SMALL_TAB_WIDTH);
  });

  resizeObserver.observe(tabNode);
  return () => resizeObserver.disconnect();
}, [ tabNode ]);

return (
  <div
    className={ classNames('tab', {
      'tab--active': active,
      'tab--dirty': dirty,
      'tab--small': small
      }) } >
      
      // ...
  </div>
); 
```

What do you think?

Reply:
LGTM